docs/configuration.md

```diff
@@ -0,0 +1,314 @@
+
+These are the configurable fields in each resource:
+
+* **KnativeServing**
+  * `spec`
+    * [config](#specconfig)
+    * [registry](#specregistry)
+      * [default](#specregistrydefault)
+      * [override](#specregistryoverride)
+      * [imagePullSecrets](#specregistryimagepullsecrets)
+    * [controller-custom-certs](#speccontroller-custom-certs)
+    * [knative-ingress-gateway](#specknative-ingress-gateway)
+    * [cluster-local-gateway](#speccluster-local-gateway)
+    * [high-availability](#spechigh-availability)
+    * [resources](#specresources)
+* **KnativeEventing**
+  * `spec`
+    * [config](#specconfig)
+    * [registry](#specregistry)
+      * [default](#specregistrydefault)
+      * [override](#specregistryoverride)
+      * [imagePullSecrets](#specregistryimagepullsecrets)
+    * [defaultBrokerClass](#specdefaultbrokerclass)
+
+
+## spec.config
+
+This is a "map of maps". The top-level keys correspond to the names of
+the Knative `ConfigMaps`, sans their "config-" prefix, e.g. the
+"config-domain" `ConfigMap` corresponds to the "domain" top-level
+key. And the map values correspond to the `data` field of each
+`ConfigMap`. 
+
+The operator will replicate this field's entries to the corresponding
+Knative `ConfigMaps`. This provides a central place to manage all your
+Knative configuration, without having to update multiple `ConfigMaps`.
+
+An example should help clarify. The following spec will cause the
+operator to update the `config-defaults`, `config-observability` and
+`config-autoscaler` `ConfigMaps` in the `knative` namespace with the
+values you see below:
+
+```
+metadata:
+  namespace: knative
+spec:
+  config:
+    defaults:
+      revision-timeout-seconds: "300"  # 5 minutes
+    observability:
+      metrics.backend-destination: prometheus
+    autoscaler:
+      stable-window: "60s"
+      container-concurrency-target-default: '100'
+```
+
+
+## spec.registry
+
+This field provides the ability to replace the images in the knative
+deployment container specs via three optional sub-fields: `override`,
+`default`, and `imagePullSecrets`.
+
+
+### spec.registry.override
+
+The optional `override` field is a mapping of deployment container
+names to docker image names.
+
+If the container names are not unique across all of your deployments,
+you can prefix the container name with the deployment name and a
+slash, e.g. `deployment/container`.
+
+Because some container specs map environment variables to image names,
+those are permitted as keys in the `override` map as well.
+
+The following example overrides only the `autoscaler` container,
+the `controller` container in the `broker` deployment spec, and the
+value of the `DISPATCHER_IMAGE` env var.
+
+```
+spec:
+  registry:
+    override:
+      autoscaler: docker.io/my-org/autoscaler:v0.13.0
+      broker/controller: docker.io/my-org/broker-controller:v0.13.0
+      DISPATCHER_IMAGE: docker.io/my-org/dispatcher:v0.13.0
```

Comment:
[not a comment on the doc]

Oh boy. This is confusing :(

Not really related to how the documentation is. Documentation is good.
The image override mechanism is confusing. 3 different patterns.

We should've accepted @pmorie 's suggestion of creating typed API for these

Reply:
I agree, fwiw, but I also think the `registry` field is an advanced use case.